### PR TITLE
Plain javascript api to asynchronously load javascript or css files

### PR DESCRIPTION
**Pull request by juliusknorr**
This PR adds a javascript only public API to replace OC.addScript and OC.addStyle (which require jquery) to dynamically load js and css files after the initial loading.

Waiting for https://github.com/nextcloud/server/pull/13156 to integrate it into the webpack build.

Uses:
- [ ] firstrunwizard: Load only when needed https://github.com/nextcloud/firstrunwizard/pull/106
- [ ] richdocuments: Replace use of OC.addScript
- [ ] files_texteditor: Replace use of OC.addScript

### CODE REVIEW

File: core/js/loader.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	 * @param {string} file
+	 */
+	loadStylesheet: function(app, file) {
+		return new Promise(function (resolve, reject) {
+			var stylePath = OC.filePath(app, 'css', file);
+			var link = document.createElement('link');
+			link.href = stylePath;
+			link.type = 'text/css';
+			link.rel = 'stylesheet';
+			link.onload = () => resolve();
```
Comment (skjnldsv) [inline core/js/loader.js]: Arrow functions not compatible with ie11 :)

**Comment by rullzer** [comment]
> Waiting for #13156 to integrate it into the webpack build.

go go go :rocket:

File: core/js/loader.js (hunk 1 above)
Comment (go2sh) [reply core/js/loader.js]: Promise is also not compatible with ie11

File: core/js/loader.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ */
+
+/**
+ * @namespace OCP
+ * @class Loader
+ */
+OCP.Loader = {
+	loadedScripts: {},
+	loadedStylesheets: {},
```
Comment (danielkesselberg) [inline core/js/loader.js]: `loadedStylesheets` is not used

File: core/js/loader.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	 *
+	 * @param {string} app
+	 * @param {string} file
+	 * @returns {Promise}
+	 */
+	loadScript: function(app, file) {
+		const key = app + file;
+		if (this.loadedScripts.hasOwnProperty(key)) {
+			return Promise.resolve();
+		} else {
```
Comment (danielkesselberg) [inline core/js/loader.js]: You could omit the else block here

File: core/js/loader.js (hunk 1 above)
Comment (juliushaertl) [reply core/js/loader.js]: @go2sh We have a polyfill for Promises :wink: 

Also this is now included in the webpack bundle so arrow functions are fine now.

**Comment by juliushaertl** [comment]
@ChristophWurst @skjnldsv I have moved the loader to the webpack build in core/ and made it extend the existing window.OCP object. Should we have this as a separate entry file or keep it in the one bundle for now? 

Besides that, it is ready for review.

File: core/js/js.js [hunk 4]
```diff
@@ -39,7 +39,7 @@ function escapeHTML(s) {
 }
 
 /** @namespace OCP */
-var OCP = {},
+var OCP = Object.assign({}, window.OCP),
```
Comment (ChristophWurst) [inline core/js/js.js]: Does not work on IE according to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

File: core/src/OCP/loader.js [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/**
+ * @namespace OCP
+ * @class Loader
+ */
+export default {
+	loadedScripts: {},
```
Comment (ChristophWurst) [inline core/src/OCP/loader.js]: do we need this globally accessible? Otherwise I would just make that a local var in the module body

**Comment by ChristophWurst** [comment]
> @ChristophWurst @skjnldsv I have moved the loader to the webpack build in core/ and made it extend the existing window.OCP object. Should we have this as a separate entrypoint and chunk or keep it in the one bundle for now?

I think that is fine for now. We should be able to refactor that and move code around if necessary :)

File: core/js/js.js (hunk 4 above)
Comment (juliushaertl) [reply core/js/js.js]: Yep, works fine.

Comment (juliushaertl) [reply core/js/js.js]: Thats what we have the polyfill for.

File: core/src/OCP/loader.js (hunk 5 above)
Comment (juliushaertl) [reply core/src/OCP/loader.js]: Fixed.

**Comment by rullzer** [comment]
> > @ChristophWurst @skjnldsv I have moved the loader to the webpack build in core/ and made it extend the existing window.OCP object. Should we have this as a separate entrypoint and chunk or keep it in the one bundle for now?
> 
> I think that is fine for now. We should be able to refactor that and move code around if necessary :)

I agree. Basically the OCP stuff should always be loaded if possible anyway. So makes sense to have this in the dist pack.

File: core/src/OCP/loader.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+			script.onerror = () => reject(`Failed to load script from ${scriptPath}`);
+			document.head.appendChild(script);
+		});
+	},
+
+	/**
+	 * Load a stylesheet file asynchronously
+	 * @param {string} app
+	 * @param {string} file
+	 */
```
Comment (danielkesselberg) [inline core/src/OCP/loader.js]: `@returns {Promise}`

Comment (skjnldsv) [reply core/src/OCP/loader.js]: Maybe we should use eslint and start enforcing valid jshint documentation?

Comment (juliushaertl) [reply core/src/OCP/loader.js]: @skjnldsv Yep, let's add that in a separate pull request. :wink:

**Comment by rullzer** [comment]
So this can go in?

**Comment by juliushaertl** [comment]
Yes, please review.

**Review by rullzer — APPROVED** [review]
Looks good to me and will make lazy loading much better :D

**Review by ChristophWurst — APPROVED** [review]
Code looks good!

**Comment by ChristophWurst** [comment]
This will give a nice conflict with https://github.com/nextcloud/server/pull/13635 🤐

File: core/js/js.js [hunk 7]
```diff
@@ -1352,7 +1354,9 @@ function initCore() {
 
 	// css variables fallback for IE
 	if (msie > 0 || trident > 0) {
-		cssVars();
+		cssVars({
+			watch: true
```
Comment (skjnldsv) [inline core/js/js.js]: poor browser will die of exhaustion :coffin:

**Review by skjnldsv — APPROVED** [review]

**Comment by MorrisJobke** [comment]
Please rebase and rebuild to resolve the conflict.

File: core/src/main.js [hunk 8]
```diff
@@ -22,3 +22,5 @@
 import '@babel/polyfill'
 
 import './globals'
+
+import './OCP/index'
```
Comment (rullzer) [inline core/src/main.js]: index.js doesn't exist ;)

**Review by rullzer — DISMISSED** [review]

File: core/src/main.js (hunk 8 above)
Comment (juliushaertl) [reply core/src/main.js]: Pushed and updated the bundled files.

**Pull request merged**